extends Node2D

#region

## SIGNALS ##
signal left_mouse_button_clicked
signal left_mouse_button_released

## CONSTANT VARIABLES ##
const COLLISION_MASK_CARD = 1
const COLLISION_MASK_DECK = 4

## STANDARD VARIABLES ##

## ON READY VARIABLES ##
@onready var card_manager: CardManager = $"../card_manager"
@onready var deck_reference: Node2D = $"../deck"

#endregion

#func _ready() -> void:
	#card_manager_reference = card_manager
	#deck_reference = deck

func _input(event: InputEvent) -> void:
	if event is InputEventMouseButton and event.button_index == MOUSE_BUTTON_LEFT :
		if event.is_pressed() :
			emit_signal( "left_mouse_button_clicked" )
			#left_mouse_button_clicked.emit()
			raycast_at_cursor()
		else : 
			emit_signal( "left_mouse_button_released" )
			#left_mouse_button_released.emit()
			pass

func raycast_at_cursor() :
	var space_state = get_world_2d().direct_space_state
	var parameters = PhysicsPointQueryParameters2D.new()
	parameters.position = get_global_mouse_position()
	parameters.collide_with_areas = true
	var result = space_state.intersect_point( parameters )
	if result.size() > 0 :
		var result_collision_mask = result[0].collider.collision_mask
		if result_collision_mask == COLLISION_MASK_CARD :
			# Card clicked
			var card_found = result[0].collider.get_parent()
			if card_found :
				card_manager.start_drag( card_found )
		elif result_collision_mask == COLLISION_MASK_DECK : 
			# Deck clicked
			deck_reference.draw_card()
			
